data/tutorials/tt_02_file_manipulation.md

```diff
@@ -9,123 +9,618 @@ date: 2021-05-27T21:07:30-00:00
+
+**Step 2: Write to the Channel**
+
+```ocaml
+Out_channel.output_string oc "Hello, World!\n"
+```
+
+Next, we write to the `out_channel` using the function
+`Out_channel.output_string`. This function takes two parameters: the
+`out_channel` to write to, and the string to write.
+
+In this case, the string `"Hello, World!\n"` is written to the file through the
+`oc` output channel.
+
+**Step 3: Flush the Channel**
+
+```ocaml
+Out_channel.flush oc
+```
+
+To make sure that the data we've written to the `out_channel` is immediately
+sent to the file, we flush the channel using `Out_channel.flush`. This function
+takes the `out_channel` as a parameter and ensures that all buffered data is
+written to the file.
+
+Flushing the channel isn't strictly necessary in this case, since the channel is
+flushed on closing. Internally, a channel is a buffer, it keeps its data in
+memory and writes to the disk when it is full. Flushing is usually necessary
+when multiple processes or threads access the channel in parallel.
+
+**Step 4: Close the Channel**
+
+```ocaml
+Out_channel.close oc
+```
+
+Finally, after we're done writing to the file, we close the `out_channel` using
+`Out_channel.close`. This function takes the `out_channel` as a parameter,
+flushes any remaining buffered data to the file, and then closes the channel.
+Closing the channel when you're done with it is crucial to free up system
+resources.
+
+## Reading from a File
+
+Here's a complete example of reading from a file in OCaml:
+
+```ocaml
+let read_from_file () =
+  (* Step 1: Open the File *)
+  let ic = In_channel.open_text "file.txt" in
+  (* Step 2: Read from the Channel *)
+  let content = In_channel.input_all ic in
+  (* Step 3: Close the Channel *)
+  In_channel.close ic;
+  content
+```
+
+Note that in practice, this is a shorter way to read all text content from a
+file:
+
+```ocaml
+let content = In_channel.with_open_text "file.txt" In_channel.input_all
+```
+
+Let's examine each step one by one.
+
+
+**Step 1: Open the File**
+
+```ocaml
+let ic = In_channel.open_text "file.txt"
+```
+
+The function`In_channel.open_text` is used to create an `in_channel` for
+reading.
+
+Here, the file `"file.txt"` is opened for reading and the returned `in_channel`
+is stored in the variable `ic`.
+
+**Step 2: Read from the Channel**
+
+```ocaml
+let content = In_channel.input_all ic
+```
+
+Next, we read from the `in_channel` using the function `In_channel.input_all`.
+This function takes the `in_channel` as a parameter and reads the entire content
+of the channel, and stores the read content in the variable `content`.
+
+**Step 3: Close the Channel**
+
+
+```ocaml
+In_channel.close ic
+```
+
+After we're done reading from the file, we close the `in_channel` using
+`In_channel.close`. This function takes the `in_channel` as a parameter and
+closes it. Similar to the `out_channel`, closing the channel is important for
+freeing up system resources.
+
+## Filesystem Operations
```

Comment:
Part of these are rarely used (`chdir`, `chmod`, etc..) and would qualify as "Advanced topics". I suggest to move this section into "Advanced topics" and to bring `Filename` handling and line-by-line reading at this position.